### PR TITLE
🌱 clusterctl: always run crd migration if possible to reduce conversion webhook usage

### PR DESCRIPTION
**Pull request by chrischdi**
**What this PR does / why we need it**:

If objects are still stored in the old version, every time an object is requested
the kube-apiserver will call the conversion webhook. Instead of lots of conversion
webhook calls we once do the migration if possible.

Also increases the timeout used for the migrations: Before running CRD migration
clusterctl usually upgrades cert-manager which may lead to updating Certificates and
rolling out new certificates to our controllers. There is a chance that this can take
up to 90s in which conversion, validating or mutatingwebhooks may be unavailable.
Because the migration gets run more aggresively during upgrades this also increases
the relevant timeouts.

---

This also fixes to not run the CRD version migration, if the stored version is equal to the storageVersion (which would result in running migration, but being a no-op because of no conversion happening).

This could e.g. be the case when upgrading from v0.4 to v1.6 or v1.7: What would happen with the old code:
* old CRD is deployed, having `storageVersion=v1alpha4` and `storedVersions = [v1alpha4]`
* `clusterctl upgrade apply`
  * CRD migration starts
    * v1alpha4 is not served anymore so it runs migration
    * migration is no-op, because storageVersion is still v1alpha4
  * CRD's get patched/updated
  * CRD's would have `storedVersions = [v1alpha4, v1beta1] (as soon as the first object got applied`

It is totally fine that v1alpha4 is not served anymore with the new CRD, important is that it is still in the versions of the new CRD, so kube-apiserver is still able to run the conversion webhook. The next upgrade would cover removing the `valpha4` from storedVersions, because the migration would be done.

We have a safeguard to not run in the error case above the changed code:

https://github.com/kubernetes-sigs/cluster-api/blob/03889674070653b0183f878ef2822646fa461905/cmd/clusterctl/client/cluster/crd_migration.go#L114

**Which issue(s) this PR fixes** *(optional, in `fixes #<issue number>(, fixes #<issue_number>, ...)` format, will close the issue(s) when PR gets merged)*:
Fixes #

/area clusterctl

### CODE REVIEW

**Comment by chrischdi** [comment]
/hold

I need to figure out the case:

```
	if !newVersions.Has(currentStorageVersion) {
```

**Comment by chrischdi** [comment]
/hold

For getting a good review

File: cmd/clusterctl/client/cluster/crd_migration.go [hunk 1]
```diff
@@ -230,3 +231,20 @@ func storageVersionForCRD(crd *apiextensionsv1.CustomResourceDefinition) (string
...
 	return "", errors.Errorf("could not find storage version for CRD %q", crd.Name)
 }
+
+// newCRDMigrationBackoff creates a new API Machinery backoff parameter set suitable for use with crd migration operations.
+// Clusterctl upgrades cert-manager right before doing CRD migration. This may lead to rollout of new certificates.
+// The time between new certificate creation + injection into objects (CRD, Webhooks) and the new secrets getting propagated
+// to the controller can be 60-90s, because the kubelet only periodically syncs secret contents to pods.
+// During this timespan conversion, validating- or mutationwebhooks may be unavailable and cause a failure.
+func newCRDMigrationBackoff() wait.Backoff {
+	// Return a exponential backoff configuration which returns durations for a total time of ~1m30s.
```
Comment (sbueringer) [inline cmd/clusterctl/client/cluster/crd_migration.go:241]: Not opposed to it but the next line seems to sum up to a lot more than 90s?

File: cmd/clusterctl/client/cluster/crd_migration.go [hunk 2]
```diff
@@ -167,7 +168,7 @@ func (m *crdMigrator) migrateResourcesForCRD(ctx context.Context, crd *apiextens
 			obj := list.Items[i]
 
 			log.V(5).Info("Migrating", logf.UnstructuredToValues(obj)...)
-			if err := retryWithExponentialBackoff(ctx, newWriteBackoff(), func(ctx context.Context) error {
+			if err := retryWithExponentialBackoff(ctx, newCRDMigrationBackoff(), func(ctx context.Context) error {
 				return handleMigrateErr(m.Client.Update(ctx, &obj))
 			}); err != nil {
```
Comment (sbueringer) [inline cmd/clusterctl/client/cluster/crd_migration.go:173]: I think the following doesn't make sense anymore: https://github.com/kubernetes-sigs/cluster-api/blob/d754ad5df262f1c5054c0582c29d9bf377452f04/cmd/clusterctl/client/cluster/crd_migration.go#L133-L135

Please check & update if necessary. I would assume that servedVersions can be dropped

**Comment by sbueringer** [comment]
@killianmuldoon PTAL. I think our whole logic around that we have to go through conversion if an apiVersion becomes unserved was flawed. A version doesn't have to be served to migrate it: a) the get & update calls can be done against the current storage version b) served is not required for conversion (apiserver will just call the conversion webhook and that's it)

File: cmd/clusterctl/client/cluster/crd_migration.go (hunk 1 above)
Comment (chrischdi) [reply cmd/clusterctl/client/cluster/crd_migration.go:241]: You should not sum it up, it are the points in time when the retry would happen.

Comment (chrischdi) [reply cmd/clusterctl/client/cluster/crd_migration.go:241]: But ack, I will take a look that we increase some more, didn't think about the jitter, which maybe could make it to less than 90s!

Comment (chrischdi) [reply cmd/clusterctl/client/cluster/crd_migration.go:241]: You are right, it does not :D

**Review by killianmuldoon — COMMENTED** [review]
Looks good - pending fixups - this is definitely easier to reason about than the previous migrations

File: cmd/clusterctl/client/cluster/crd_migration.go [hunk 3]
```diff
@@ -129,9 +126,8 @@ func (m *crdMigrator) run(ctx context.Context, newCRD *apiextensionsv1.CustomRes
 	// This way we can make sure that all CR objects are now stored in the current storage version.
```
Comment (sbueringer) [inline cmd/clusterctl/client/cluster/crd_migration.go:126]: Some parts of this godoc seem oudated.

I would suggest:
* remove l.122+123
* move l.124 to the end of the godoc block

File: cmd/clusterctl/client/cluster/crd_migration.go [hunk 4]
```diff
@@ -230,3 +226,20 @@ func storageVersionForCRD(crd *apiextensionsv1.CustomResourceDefinition) (string
 	}
 	return "", errors.Errorf("could not find storage version for CRD %q", crd.Name)
 }
+
+// newCRDMigrationBackoff creates a new API Machinery backoff parameter set suitable for use with crd migration operations.
+// Clusterctl upgrades cert-manager right before doing CRD migration. This may lead to rollout of new certificates.
+// The time between new certificate creation + injection into objects (CRD, Webhooks) and the new secrets getting propagated
+// to the controller can be 60-90s, because the kubelet only periodically syncs secret contents to pods.
+// During this timespan conversion, validating- or mutationwebhooks may be unavailable and cause a failure.
```
Comment (sbueringer) [inline cmd/clusterctl/client/cluster/crd_migration.go:234]:
```suggestion
// During this timespan conversion, validating- or mutating-webhooks may be unavailable and cause a failure.
```
nit: or something like that :D

File: cmd/clusterctl/client/cluster/crd_migration.go [hunk 5]
```diff
@@ -230,3 +226,20 @@ func storageVersionForCRD(crd *apiextensionsv1.CustomResourceDefinition) (string
...
+func newCRDMigrationBackoff() wait.Backoff {
+	// Return a exponential backoff configuration which returns durations for a total time of ~1m30s.
```
Comment (sbueringer) [inline cmd/clusterctl/client/cluster/crd_migration.go:236]: time of 1m30s + some buffer?

File: cmd/clusterctl/client/cluster/crd_migration_test.go [hunk 6]
```diff
@@ -111,6 +111,28 @@ func Test_CRDMigrator(t *testing.T) {
...
+					Versions: []apiextensionsv1.CustomResourceDefinitionVersion{
+						{Name: "v1beta1", Storage: true, Served: true}, // v1beta1 is being added
+						{Name: "v1alpha1", Served: false},              // v1alpha1 still exists but is not served anymore
+					},
+				},
+			},
+			wantIsMigrated: false,
+		},
 		{
 			name: "Fails if new CRD drops current storage version",
```
Comment (sbueringer) [inline cmd/clusterctl/client/cluster/crd_migration_test.go:137]: Can we simplify the test cases? There is a lot of misleading language in the test names now, e.g. "Migrate CRs if their storage version is removed from the CRD" (that we migrate is now solely determined by the comparison of the status field and the storage version)

**Comment by sbueringer** [comment]
Thx!

/approve

**Comment by sbueringer** [comment]
/hold cancel

Assuming a good review has been provided :joy:

**Pull request merged**